### PR TITLE
feat: add Object closeOnResolve option

### PR DESCRIPTION
**Pull request by toomuchdesign**
This PR adds a `closeOnResolve` boolean option to `Object` type allowing ignoring existing `openProps` option on resolve.

See: https://github.com/ThomasAribart/json-schema-to-ts/issues/108#issuecomment-1987288210

### CODE REVIEW

**Comment by toomuchdesign** [comment]
Hi @ThomasAribart!
This is my first contribution attempt to this fascinating library!

I noticed that `closeOnResolve` and `openProps` options seem to do quite the opposite.

I'm definitely missing how `ts-algebra` is used in `json-schema-to-ts` therefore I assumed there's a reason why we cannot simply set `openProps` to `Never` (maybe `openProps` and `closeOnResolve` are known in different execution times?)

I decided to strictly follow your instructions and use `closeOnResolve` just at resolve time keeping both object's internal props:
- `isOpen`
- `closeOnResolve`

Happy to rework the PR based on your feedback!
Thank you

**Comment by toomuchdesign** [comment]
I've just noticed there's more to update/refactor!
@ThomasAribart please hang on with the review until the PR is ready :)

File: src/meta-types/intersection/object.ts [hunk 1]
```diff
@@ -121,12 +128,17 @@ type IntersectObjects<
     ObjectOpenProps<META_OBJECT_A>,
     ObjectOpenProps<META_OBJECT_B>
   >,
+  INTERSECTED_CLOSE_ON_RESOLVE = And<
```
Comment (toomuchdesign) [inline src/meta-types/intersection/object.ts]: Intersection: is this the expected behaviour for the `closeOnResolve` option?

File: src/meta-types/object.ts [hunk 2]
```diff
@@ -152,13 +172,17 @@ export type ResolveObject<
...
+      ObjectCloseOnResolve<META_OBJECT>,
+      {},
       If<
-        IsObjectEmpty<META_OBJECT>,
-        { [KEY: string]: Resolve<ObjectOpenProps<META_OBJECT>, OPTIONS> },
-        { [KEY: string]: Resolve<Any, OPTIONS> }
-      >,
-      {}
+        IsObjectOpen<META_OBJECT>,
+        If<
```
Comment (toomuchdesign) [inline src/meta-types/object.ts]: I handled `closeOnResolve` and `openProps` independently until type resolution.

If viable, another approach could have consisted of setting `openProps` to `Never` when `closeOnResolve` is true.

**Comment by toomuchdesign** [comment]
Hi again @ThomasAribart! Tests are now all green and the PR good enough for a first round of review. I left a couple comments to highlight a couple debatable implementation decisions.

File: src/meta-types/intersection/object.ts (hunk 1 above)
Comment (ThomasAribart) [reply src/meta-types/intersection/object.ts]: It made more sense to use a `Or` instead :)

**Comment by ThomasAribart** [comment]
Hi @toomuchdesign and many thanks for this PR! I tweaked some files but it was minor, great job!

`openProps` and `closeOnResolve` are two different options and it makes sense to keep them that way. The former will by applied when intersecting two object schemas to check for compatibility. The latter is more of a "cosmetic" choice that applies when resolving.

**Review by ThomasAribart — APPROVED** [review]

**Pull request merged**